### PR TITLE
[BANDWIDTH_THROTTLING] Update QuotaManager interface and introduce AtomicAmbryQuotaManager.

### PR DESCRIPTION
**Pull request by ankagrawal**

### CODE REVIEW

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java [hunk 1]
```diff
@@ -77,86 +88,138 @@ public AmbryQuotaManager(QuotaConfig quotaConfig, QuotaRecommendationMergePolicy
...
+   * and makes a best effort to gather as many recommendations as possible. Recommendations for some {@link QuotaName}s
+   * will be missing in those cases.
+   */
   @Override
-  public ThrottlingRecommendation getThrottleRecommendation(RestRequest restRequest) {
-    if (requestQuotaEnforcers.isEmpty()) {
+  public ThrottlingRecommendation recommend(RestRequest restRequest) throws QuotaException {
+    quotaMetrics.recommendRate.mark();
+    if (quotaEnforcers.isEmpty()) {
+      LOGGER.warn("No quota enforcers in AmbryQuotaManager. Returning null recommendation.");
```
Comment (justinlin-linkedin) [inline ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:122]: let's use trace or debug here, but add a warn log in the constructor. This is probably not going to happen, but if it does, we are going to see tons of warn log messages in the file.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java [hunk 2]
```diff
@@ -77,86 +88,138 @@ public AmbryQuotaManager(QuotaConfig quotaConfig, QuotaRecommendationMergePolicy
...
+   * {@link QuotaEnforcer}s as possible and send a {@link QuotaAction#ALLOW} recommendation. If there is an exception in
+   * any {@link QuotaEnforcer#charge} call, that exception is logged and charge is still attempted for rest of the
+   * {@link QuotaEnforcer}s. As such, there is a potential for partial charge to happen in this implementation.
+   */
   @Override
-  public ThrottlingRecommendation charge(RestRequest restRequest, BlobInfo blobInfo,
-      Map<QuotaName, Double> requestCostMap) {
-    if (requestQuotaEnforcers.isEmpty()) {
-      return null;
+  public QuotaAction chargeAndRecommend(RestRequest restRequest, BlobInfo blobInfo,
```
Comment (justinlin-linkedin) [inline ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:172]: doesn't seem like that BlobInfo is used anywhere, now that we are changing the interface, we might as well just remove the BlobInfo once for all.

**Review by justinlin-linkedin — COMMENTED** [review]
most code looks good to me

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java [hunk 3]
```diff
@@ -77,86 +88,138 @@ public AmbryQuotaManager(QuotaConfig quotaConfig, QuotaRecommendationMergePolicy
...
       return null;
     }
     ThrottlingRecommendation throttlingRecommendation;
-    Timer.Context timer = quotaMetrics.quotaEnforcementTime.time();
+    Timer.Context timer = quotaMetrics.quotaRecommendationTime.time();
     try {
       List<QuotaRecommendation> quotaRecommendations = new ArrayList<>();
-      for (QuotaEnforcer quotaEnforcer : requestQuotaEnforcers) {
+      boolean partialRecommendation = false;
+      boolean noRecommendation = true;
```
Comment (tbingol) [inline ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:131]: Nit: These can be inferred from the size of quotaRecommendations

quotaRecommendations.empty() => no recommendation
quotaRecommendations.size() < quotaEnforcers.size() => partial recommendation

**Review by tbingol — APPROVED** [review]
Please add a description to the PR

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java (hunk 1 above)
Comment (ankagrawal) [reply ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:122]: done.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java (hunk 3 above)
Comment (ankagrawal) [reply ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:131]: fixed.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java (hunk 2 above)
Comment (ankagrawal) [reply ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:172]: done.

**Pull request merged**